File: apps/zxtune-android/zxtune/src/androidTest/java/app/zxtune/io/IoTest.java
package app.zxtune.io;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;

import org.junit.Test;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.nio.ByteBuffer;
import java.nio.MappedByteBuffer;

public class IoTest {

  private static void generate(int fill, int size, File obj) throws IOException {
    final FileOutputStream out = new FileOutputStream(obj);
    for (int off = 0; off < size; ++off) {
      out.write(fill);
    }
    out.close();
  }

  private static File createFile(int fill, int size) throws IOException {
    final File result = File.createTempFile("test", "io");
    if (size != 0) {
      generate(fill, size, result);
    }
    assertTrue(result.canRead());
    assertEquals(size, result.length());
    return result;
  }

  private static void checkBuffer(ByteBuffer buf, int fill, int size) {
    assertEquals(size, buf.limit());
    for (int idx = 0; idx < size; ++idx) {
      assertEquals(fill, buf.get(idx));
    }
  }

  @Test
  public void testNonexisting() {
    final File file = new File("nonexisting_name");
    assertFalse(file.exists());
    try {
      Io.readFrom(file);
      fail("Unreachable");
    } catch (IOException e) {
      assertNotNull("Thrown exception", e);
    }
    try {
      //really FileInputStream ctor throws
      Io.readFrom(new FileInputStream(file));
      fail("Unreachable");
    } catch (IOException e) {
      assertNotNull("Thrown exception", e);
    }
  }

  @Test
  public void testEmpty() throws IOException {
    final File file = createFile(0, 0);
    try {
      Io.readFrom(file);
      fail("Unreachable");
    } catch (IOException e) {
      assertNotNull("Thrown exception", e);
    }
    final InputStream stream = new FileInputStream(file);
    try {
      Io.readFrom(stream);
      fail("Unreachable");
    } catch (IOException e) {
      assertNotNull("Thrown exception", e);
    }
  }

  @Test
  public void test1k() throws IOException {
    final int size = 1024;
    assertTrue(size < Io.INITIAL_BUFFER_SIZE);
    assertTrue(size < Io.MIN_MMAPED_FILE_SIZE);
    final File file = createFile(1, size);
    {
      final ByteBuffer buf = Io.readFrom(file);
      assertTrue(buf.isDirect());
      // Seems like for some versions DirectBuffer is mapped really
      //assertFalse(buf instanceof MappedByteBuffer);
      checkBuffer(buf, 1, size);
    }
    {
      final ByteBuffer buf = Io.readFrom(new FileInputStream(file));
      assertFalse(buf.isDirect());
      checkBuffer(buf, 1, size);
    }
    {
      final ByteBuffer buf = Io.readFrom(new FileInputStream(file), size);
      assertTrue(buf.isDirect());
      checkBuffer(buf, 1, size);
    }
  }

  @Test
  public void test300k() throws IOException {
    final int size = 300 * 1024;
    assertTrue(size > Io.INITIAL_BUFFER_SIZE);
    assertTrue(size > Io.MIN_MMAPED_FILE_SIZE);
    final File file = createFile(2, size);
    {
      final ByteBuffer buf = Io.readFrom(file);
      assertTrue(buf.isDirect());
      assertTrue(buf instanceof MappedByteBuffer);
      checkBuffer(buf, 2, size);
    }
    {
      final ByteBuffer buf = Io.readFrom(new FileInputStream(file));
      assertFalse(buf.isDirect());
      checkBuffer(buf, 2, size);
    }
    {
      final ByteBuffer buf = Io.readFrom(new FileInputStream(file), size);
      assertTrue(buf.isDirect());
      checkBuffer(buf, 2, size);
    }
  }

  @Test
  public void testWrongSizeHint() throws IOException {
    final File file = createFile(3, 2 * 1024);
    checkBuffer(Io.readFrom(new FileInputStream(file)), 3, 2 * 1024);
    try {
      Io.readFrom(new FileInputStream(file), 1024);
      fail("Unreachable");
    } catch (IOException e) {
      assertNotNull("Thrown exception", e);
    }
    try {
      Io.readFrom(new FileInputStream(file), 3 * 1024);
      fail("Unreachable");
    } catch (IOException e) {
      assertNotNull("Thrown exception", e);
    }
  }
}

File: apps/zxtune-android/zxtune/src/androidTest/java/app/zxtune/io/IoTest.kt
package app.zxtune.io

import org.junit.Assert.*
import org.junit.Test
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException
import java.nio.ByteBuffer
import java.nio.MappedByteBuffer

class IoTest {
    @Test
    fun testNonexisting() = File("nonexisting_name").let { file ->
        assertFalse(file.exists())
        expectThrowsIOException { Io.readFrom(file) }
        expectThrowsIOException {
            //really FileInputStream ctor throws
            Io.readFrom(FileInputStream(file))
        }
    }

    @Test
    fun testEmpty() = createFile(0, 0).let { file ->
        expectThrowsIOException { Io.readFrom(file) }
        FileInputStream(file).let { stream ->
            expectThrowsIOException { Io.readFrom(stream) }
        }
    }

    @Test
    fun test1k() {
        val size = 1024
        assertTrue(size < Io.INITIAL_BUFFER_SIZE)
        assertTrue(size < Io.MIN_MMAPED_FILE_SIZE)
        val file = createFile(1, size)
        Io.readFrom(file).let { buf ->
            assertTrue(buf.isDirect)
            // Seems like for some versions DirectBuffer is mapped really
            //assertFalse(buf is MappedByteBuffer)
            checkBuffer(buf, 1, size)
        }
        Io.readFrom(FileInputStream(file)).let { buf ->
            assertFalse(buf.isDirect)
            checkBuffer(buf, 1, size)
        }
        Io.readFrom(FileInputStream(file), size.toLong()).let { buf ->
            assertTrue(buf.isDirect)
            checkBuffer(buf, 1, size)
        }
    }

    @Test
    fun test300k() {
        val size = 300 * 1024
        assertTrue(size > Io.INITIAL_BUFFER_SIZE)
        assertTrue(size > Io.MIN_MMAPED_FILE_SIZE)
        val file = createFile(2, size)
        Io.readFrom(file).let { buf ->
            assertTrue(buf.isDirect)
            assertTrue(buf is MappedByteBuffer)
            checkBuffer(buf, 2, size)
        }
        Io.readFrom(FileInputStream(file)).let { buf ->
            assertFalse(buf.isDirect)
            checkBuffer(buf, 2, size)
        }
        Io.readFrom(FileInputStream(file), size.toLong()).let { buf ->
            assertTrue(buf.isDirect)
            checkBuffer(buf, 2, size)
        }
    }

    @Test
    fun testWrongSizeHint() = createFile(3, 2 * 1024).let { file ->
        checkBuffer(Io.readFrom(FileInputStream(file)), 3, 2 * 1024)
        expectThrowsIOException { Io.readFrom(FileInputStream(file), 1024) }
        expectThrowsIOException { Io.readFrom(FileInputStream(file), (3 * 1024).toLong()) }
    }
}

private fun expectThrowsIOException(cmd: () -> Unit) {
    try {
        cmd()
        fail("Unreachable")
    } catch (e: IOException) {
        assertNotNull("Thrown exception", e)
    }
}

private fun generate(fill: Int, size: Int, obj: File) = FileOutputStream(obj).use { out ->
    repeat(size) {
        out.write(fill)
    }
}

private fun createFile(fill: Int, size: Int) = File.createTempFile("test", "io").apply {
    if (size != 0) {
        generate(fill, size, this)
    }
    assertTrue(canRead())
    assertEquals(size.toLong(), length())
}

private fun checkBuffer(buf: ByteBuffer, fill: Int, size: Int) {
    assertEquals(size.toLong(), buf.limit().toLong())
    for (idx in 0 until size) {
        assertEquals(fill.toLong(), buf[idx].toLong())
    }
}
